keps/sig-cli/3659-kubectl-apply-prune/README.md

```diff
@@ -0,0 +1,1365 @@
+
+When a tool that wants to list the members of an ApplySet and cannot determine the list of GKs (i.e. because neither hint annotation is used), it may support “discovery”, likely warning that a full cluster crawl is being attempted. Insufficient permissions errors are likely with such functionality, and when they are encountered, the tool should warn that the membership list may be incomplete.
+
+```yaml
+applyset.k8s.io/contains-group-kinds: <kind>.<group>[,]` # OPTIONAL
+```
+
+The `applyset.k8s.io/contains-group-kinds` annotation is an optional "hint" annotation tools can populate and use to optimize listing of member objects. Tooling not using this annotation may safely ignore it. Since the annotation on the member objects themselves remains the source of truth for set membership, tools making use of this optimization should consider also providing or periodically automating a resync of the hint annotation.
+
+When present, the value of this annotation shall be a comma separated list of the group-kinds, in the fully-qualified name format, i.e. `<resourcename>.<group>`.  An example annotation value might therefore look like: `certificates.cert-manager.io,configmaps,deployments.apps,secrets,services`.  Note that we do not include the version; formatting is a different concern from “ApplySet membership”.
+
+To avoid spurious updates and conflicts, the list must be sorted alphabetically.  The list may include GKs where there are no resources actually labeled with the ApplySet-id, but to avoid churn this should be avoided and ideally only be a transitional step during ApplySet mutations.
+
+```yaml
+applyset.k8s.io/inventory: <kind>.<group>/<name>.<namespace>[,] # OPTIONAL
+```
+
+The `applyset.k8s.io/inventory` annotation is an alternative optional "hint" annotation tools can populate and use to optimize listing of member objects. Tooling not using this annotation may safely ignore it. Since the annotation on the member objects themselves remains the source of truth for set membership, tools making use of this optimization should consider also providing or periodically automating a resync of the hint annotation.
+
+When used, its value must be a comma separated list of all the GKNNs in use in the ApplySet. To avoid spurious updates and conflicts, the list must be sorted alphabetically.
+
+Tooling using this annotation should take care to ensure that all listed GKNNs are in fact valid members of the ApplySet based on its scope. For instance, a cluster-scoped parent without a `applyset.k8s.io/additional-namespaces` annotation cannot reference namespace-scoped GKNNs, and a namespace-scoped parent without that annotation cannot reference GKNNs in other namespaces.
+
+To remain compliant with the specification, tools using this particular annotation should still refrain from operating on (e.g. deleting) a member object before verifying its `applyset.k8s.io/part-of` and `applyset.k8s.io/controller-ref` annotations.
```

Comment:
I recommend for small sets, to require to store a list of objects. Lists with label selectors are extremely heavyweight for apiserver.

Reply:
Well I have confidence in team apimachinery to squeeze performance from what appears to be a brick wall :-)

But ... this is a long way from the only user of label selectors.  And in fact, we know that more sophisticated appliers than kubectl will actually start a watch on every object, so that they can report the health of the objects they apply.  (We're not opposed to contributing that logic to kubectl apply/prune, but we don't believe it is considered in scope today).  My understanding is that fewer watches are better here, but maybe you know of some benchmarks?

Now, if we had the ability to filter on arbitrary fields with arbitrary CEL expressions, or the ability to do watches across object types, that would be a different story :-)